src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,902 @@
+static uint64 ConvertLocalTableToShard(Oid relationId);
+static void RenameRelationToShardRelation(Oid shellRelationId, uint64 shardId);
+static void RenameShardRelationConstraints(Oid shardRelationId, uint64 shardId);
+static List * GetConstraintNameList(Oid relationId);
+static void RenameForeignConstraintsReferencingToShard(Oid shardRelationId,
+													   uint64 shardId);
+static char * GetRenameShardConstraintCommand(Oid relationId, char *constraintName,
+											  uint64 shardId);
+static void RenameShardRelationIndexes(Oid shardRelationId, uint64 shardId);
+static char * GetRenameShardIndexCommand(char *indexName, uint64 shardId);
+static List * GetExplicitIndexNameList(Oid relationId);
+static void CreateCitusLocalTable(Oid relationId);
+static void FinalizeCitusLocalTableCreation(Oid relationId);
+static List * GetShellTableDDLEventsForCitusLocalTable(Oid relationId);
+static void CreateShellTableForCitusLocalTable(List *shellTableDDLEvents);
+static void DropAndMoveDefaultSequenceOwnerships(Oid sourceRelationId,
+												 Oid targetRelationId);
+static void ExtractColumnsOwningSequences(Oid relationId, List **columnNameList,
+										  List **ownedSequenceIdList);
+static void DropDefaultColumnDefinition(Oid relationId, char *columnName);
+static void TransferSequenceOwnership(Oid ownedSequenceId, Oid targetRelationId,
+									  char *columnName);
+static void InsertMetadataForCitusLocalTable(Oid citusLocalTableId, uint64 shardId);
+
+
+PG_FUNCTION_INFO_V1(create_citus_local_table);
+
+
+/*
+ * create_citus_local_table creates a citus table from the table with relationId.
+ * by executing the internal method CreateCitusLocalTable.
+ * (See CreateCitusLocalTable function's comment.)
+ */
+Datum
+create_citus_local_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	Oid relationId = PG_GETARG_OID(0);
+
+	CreateCitusLocalTable(relationId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * ErrorIfUnsupportedCreateCitusLocalTable errors out if we cannot create the
+ * citus local table from the relation with relationId.
+ */
+static void
+ErrorIfUnsupportedCreateCitusLocalTable(Relation relation)
+{
+	if (!RelationIsValid(relation))
+	{
+		ereport(ERROR, (errmsg("cannot create citus local table, relation does "
+							   "not exist")));
+	}
+
+	Oid relationId = relation->rd_id;
+
+	ErrorIfCoordinatorNotAddedAsWorkerNode(relationId);
+	ErrorIfUnsupportedCitusLocalTableKind(relationId);
+	EnsureTableNotDistributed(relationId);
+
+	/*
+	 * When creating other citus table types, we don't need to check that case as
+	 * EnsureTableNotDistributed already errors out if the given relation implies
+	 * a citus table. However, as we don't mark the relation as citus table, i.e we
+	 * do not use the relation with relationId as the shell relation, parallel
+	 * create_citus_local_table executions would not error out for that relation.
+	 * Hence we need to error out for shard relations too.
+	 */
+	ErrorIfRelationIsAKnownShard(relationId);
```

Comment:
For
`ErrorIfCoordinatorNotAddedAsWorkerNode`,
`ErrorIfRelationIsAKnownShard` and
`ErrorIfTableHasExternalForeignKeys`,
I could not imagine a better way to emphasize that function errors out for citus local table creation. Maybe not emphasizing `citus local table` in error messages might work.

Does it sound good ?

Reply:
If we remove 'citus local table' from these messages, it might make sense to put these methods to a common place as non-static so that they can be used in other places as well.